### PR TITLE
Create AuthService

### LINKED ISSUE #208: Create AuthService #208
### Description
Create `AuthService`

### Requirements
- project should be runnable
- should allow registering and logging in to system without exceptions.
- should allow granting role
- test all the dependent services

### Deliverables
- `SignUpDetails`, `SignInDetails` entities
- mapping between `SignUpDetails` and `User` models
- register all necessary infrastructure to DI
- `AuthController` controller with the following actions
    - `SignUp` - to sign up a new user
    - `SignIn` - to sign in a new user
    - `GrantRole` - to grand role to a user
    - `RevokeRole` - to revoke role from a user

### PR DESCRIPTION
**Pull request by rashidovasadbek**
Closes #208

### CODE REVIEW

**Review by sulton-max — CHANGES_REQUESTED** [review]

File: src/Training.AirBnb.Clone.Backend/AirBnB.Domain/Entities/User.cs [hunk 1]
```diff
@@ -27,15 +27,10 @@ public sealed class User : AuditableEntity
...
-    ///<summary>
-    /// Gets or sets the Role Id of user
-    /// </summary>
-    public Guid RoleId { get; set; }
-
     /// <summary>
     /// Gets or sets the User role type
     /// </summary>
-    public Role Role { get; set; }
+    public IList<UserRole> Roles { get; set; }
```
[the remark this answers is not in the document]
Comment (sulton-max) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Domain/Entities/User.cs:33]: Reference - https://github.com/WoW-2-0/Backend.ILA.DataAccess/tree/main/src/EntityConfiguration.ManyToMany.DirectNavigation

**Review by Abdurasulovich — COMMENTED** [review]
Done fine

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Identity/Services/RoleProcessingService.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        var user = await userService.Get(asNoTracking: true)
+            .Include(user => user.Roles)
+            .FirstOrDefaultAsync(user => user.Id == userId,
+                cancellationToken: cancellationToken) ?? throw new InvalidOperationException("User not found");
+        
+        //Choosing the right role
+        var selectedRole = user.Roles.FirstOrDefault(role => role.Type == roleType) ?? throw new AuthenticationException(" Invalid role type");
+        
+        //Delete the selected role
+        appDbContext.Roles.Remove(selectedRole);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Identity/Services/RoleProcessingService.cs:31]: Why we are removing role from db context ? We need to remove it from user.Roles

**Review by sulton-max — CHANGES_REQUESTED** [review]

**Review by Ilxom-22 — CHANGES_REQUESTED** [review]

**Review by sulton-max — APPROVED** [review]
Looks good to me

**Review by sulton-max — APPROVED** [review]

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Identity/Models/SignUpDetails.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    /// <summary>
+    /// Gets or sets last name of the sign up details
+    /// </summary>
+    public string LastName { get; set; } = default!;
+    
+    /// <summary>
+    /// Gets or sets age of the sign up details
+    /// </summary>
+    public int Age { get; set; }
```
[the remark this answers is not in the document]
Comment (sulton-max) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Identity/Models/SignUpDetails.cs:21]: Yes, that would be actually correct, let's change that when implementing host sign up, we can close this for now

**Review by abdurakh — APPROVED** [review]
Done ✅

**Review by Ilxom-22 — APPROVED** [review]

**Pull request merged**